libs/wingsdk/src/shared-aws/bucket.ts

```diff
@@ -37,3 +45,133 @@ export class Bucket {
     );
   }
 }
+
+/**
+ * A reference to an external S3 bucket.
+ * @inflight `@winglang/sdk.cloud.IBucketClient`
+ */
+export class BucketRef extends Resource {
+  /** @internal */
+  public [INFLIGHT_SYMBOL]?: IBucketClient;
+
+  /**
+   * The Name of this bucket.
+   */
+  public readonly bucketName: string;
+
+  /**
+   * The ARN of this bucket.
+   */
+  public readonly bucketArn: string;
+
+  constructor(scope: Construct, id: string, bucketName: string) {
+    super(scope, id);
+
+    this.bucketName = bucketName;
+    this.bucketArn = `arn:aws:s3:::${bucketName}`;
+
+    this.addUserInterface();
+  }
+
+  public onLift(host: IInflightHost, ops: string[]): void {
+    // if this is an AWS function, add the necessary IAM permissions
+    const fn = Function.from(host);
+
+    if (fn) {
+      fn.addPolicyStatements(
+        ...calculateBucketPermissions(this.bucketArn, ops)
+      );
+    }
+
+    // The bucket name needs to be passed through an environment variable since
+    // it may not be resolved until deployment time.
+    host.addEnvironment(this.envName(), this.bucketName);
+
+    super.onLift(host, ops);
+  }
+
+  /** @internal */
+  public _toInflight(): string {
+    return InflightClient.for(__dirname, __filename, "BucketClient", [
+      `process.env["${this.envName()}"]`,
+    ]);
+  }
+
+  private envName(): string {
+    return `BUCKET_NAME_${this.node.addr.slice(-8)}`;
+  }
+
+  /** @internal */
+  public _supportedOps(): string[] {
+    return [
+      cloud.BucketInflightMethods.DELETE,
+      cloud.BucketInflightMethods.GET,
+      cloud.BucketInflightMethods.GET_JSON,
+      cloud.BucketInflightMethods.LIST,
+      cloud.BucketInflightMethods.PUT,
+      cloud.BucketInflightMethods.PUT_JSON,
+      cloud.BucketInflightMethods.PUBLIC_URL,
+      cloud.BucketInflightMethods.EXISTS,
+      cloud.BucketInflightMethods.TRY_GET,
+      cloud.BucketInflightMethods.TRY_GET_JSON,
+      cloud.BucketInflightMethods.TRY_DELETE,
+      cloud.BucketInflightMethods.SIGNED_URL,
+      cloud.BucketInflightMethods.METADATA,
+      cloud.BucketInflightMethods.COPY,
+      cloud.BucketInflightMethods.RENAME,
+    ];
+  }
+
+  private addUserInterface() {
+    Node.of(this).color = "amber";
+
+    const awsConsoleHandler = Testing.makeHandler(
```

Comment:
Maybe we can reuse some of the code we already have in `QueueRef` so this ugliness is at least centralized in one place?